- Clamp frame to n_frames to prevent oob errors

PardhavMaradani opened the pull request:
Fixes BradyAJohnston/MolecularNodes#812

BradyAJohnston commented:
I'm a little hesitant to hide away error messages for out of bounds completely. There should be some kind of feedback for a user as to why their trajectory might not be animating how they expect.

PardhavMaradani commented:
Hi Brady, I have added a label in the UI in the Object panel to show the number of trajectory frames. The `frame` and `frame_hidden` properties will show the actual trajectory frame for the `update_with_scene` `False` and `True` scenarios respectively.

Looking deeper, I noticed several other out of bound cases and also an error in `_update_positions` which is using the `subframes`, `average`, etc values even when the `update_with_scene` is turned off. The frame mapper also doesn't seem to have any validations, especially for the case when a mapping array is passed. I fixed the out of bounds errors in the updated commit and verified it with both MDA and oxDNA trajectories (having multiple frames) - with `Update with Scene` turned off and on with all the sub options of subframes, averaging, offset and interpolation.

BradyAJohnston commented:
Thanks for diving deeper on this. I'm happy with these changes for now. I'm a little bit still of the opinion that we should be more verbose about if a user (especially via the API) is trying to animate through frames of a trajectory but they are ultimately outside of the range of them. I suppose could just be up to the user to ensure they know about their frame range or not.

PardhavMaradani commented:
Hi Brady, I agree. For the API, it looks like there are two different entry points for changing the trajectory frame today, which are mutually exclusive (for the recursion reasons you outlined earlier): The `.frame` property and the `set_frame` method. The first one changes the bpy prop (`mn.frame`) and internally calls the `set_frame` method - this seems to be the preferred way when `update with scene` is disabled (It is easy to add validations and raise exceptions here). The second one gets called in the frame handler and seems more like an internal function. This needs to however be called for the `update with scene` enabled to support the sub frames, averaging, offset etc. There are also a couple of pending issues like updating the `mn.frame_hidden` property which could cause problems and the lack of validation in the frame mapper. So, this definitely needs a slightly larger refactoring, which might be a bit more disruptive (I am also not very clear how this could be done at this point) - something we could possibly take up later?

BradyAJohnston commented:
It's not something I'm clear on how we should be doing it either, I just know that we should be doing it better than we are currently doing it. I did spend a whole day a week ago trying to refactor it and just ended up going nowhere. Something to deal with in the future and maybe to try and do some higher-level planning beforehand.
